### PR TITLE
Add with_callback that take a Fn taking request and returning responses

### PR DESCRIPTION
**Pull request by izissise**
#458 

This compile but I haven't successfully used it yet.

It takes a Fn but should probably take a FnMut or Stream for more convenience, but I couldn't make it work with RefCell or RWLock.

 Any help is appreciated, especially if you known your way around Send + Sync + Async Stream :)

### CODE REVIEW

**Comment by izissise** [comment]
I have a working version and an example use case which mod the request. I'll update this branch soon

**Comment by izissise** [comment]
I don't known how to write the example though

File: src/docker.rs [hunk 1]
```diff
@@ -589,6 +601,86 @@ impl Docker {
...
+            .next()
+            .map(|s| s.to_owned())
+            .unwrap_or_default();
+        let client_addr = client_addr
+            .trim_start_matches("unix://")
+            .trim_start_matches("npipe://")
+            .trim_start_matches("http://")
+            .trim_start_matches("https://")
+            .trim_start_matches("tcp://")
+            .to_owned();
```
Comment (fussybeaver) [inline src/docker.rs:668]: This PR looks really good.. I'm wondering here about the `client_addr`, could you unpack it from the `split` that's done in the line above, or remove the characters leading up to `://` ?

**Comment by fussybeaver** [comment]
This looks great, I'm going to ask the others who wanted this to see if they need anything else...

File: src/docker.rs [hunk 2]
```diff
@@ -90,8 +90,16 @@ pub(crate) enum ClientType {
...
+    Callback {
+        scheme: String,
+    },
 }
 
+pub type CallbackRetTy =
+    Pin<Box<dyn Future<Output = Result<Response<hyper::body::Incoming>, Error>> + Send>>;
+/// The type of the callback for custom transport.
+pub type CallbackTy = Box<dyn Fn(Request<BodyType>) -> CallbackRetTy + Send + Sync>;
+
```
Comment (DDtKey) [inline src/docker.rs:102]: For a more ergonomic API I would suggest using a `trait` instead of the `Fn`, but with default impl for `Fn`

Something like: 

```rs
trait CustomTransport {
  fn request(request: Request<BodyType>) -> ...
}

impl<CallbackTy> CustomTransport for CallbackTy 
where ...
{
    ...
}
```

This way, users can create their own (isolated) structures which may simplify the interaction with API and the same time be more flexible for extensions

WDYT?

**Review by DDtKey — COMMENTED** [review]
I really appreciate the work done here! ❤️ 

I left only one subjective suggestion, if you find it valuable

File: src/docker.rs (hunk 1 above)
Comment (izissise) [reply src/docker.rs]: I changed this to use split_once, tell me if this is what you wanted

File: src/docker.rs (hunk 2 above)
Comment (izissise) [reply src/docker.rs]: I added the trait :)

Comment (izissise) [reply src/docker.rs]: Hmm I don't known how to make it work with closures

File: src/docker.rs [hunk 3]
```diff
@@ -1317,3 +1426,6 @@ pub(crate) type BodyType = http_body_util::Either<
 pub(crate) fn body_stream(body: impl Stream<Item = Bytes> + Send + 'static) -> BodyType {
     BodyType::Right(StreamBody::new(Box::pin(body.map(|a| Ok(Frame::data(a))))))
 }
+
+/// A request from bollard used in callbacks
+pub type BollardRequest = Request<BodyType>;
```
Comment (fussybeaver) [inline src/docker.rs:1431]: I'm happy where this is going... I think we can merge soon.

Just a nitpic - could we remove the alias and use the full type where it is used (e.g. the documentation). This makes it clearer what it is at the expense of verbosity.

File: src/docker.rs (hunk 2 above)
Comment (DDtKey) [reply src/docker.rs]: I've left two suggestions which makes it working

Comment (DDtKey) [reply src/docker.rs]: To make it simpler to check the diff I've opened a PR to your branch: https://github.com/izissise/bollard/pull/1/files

File: src/docker.rs (hunk 1 above)
Comment (fussybeaver) [reply src/docker.rs]: Looks great!

File: src/docker.rs [hunk 4]
```diff
@@ -589,6 +614,78 @@ impl Docker {
...
+    ///     Box::new(move |req: BollardRequest| {
+    ///         let client = std::sync::Arc::clone(&client);
+    ///         Box::pin(async move {
+    ///             let (p, b) = req.into_parts();
+    ///             // let _prev = p.headers.insert("host", host);
+    ///             // let mut uri = p.uri.into_parts();
+    ///             //uri.path_and_query = uri.path_and_query.map(|paq|
+    ///             //   uri::PathAndQuery::try_from("/docker".to_owned() + paq.as_str())
+    ///             // ).transpose().map_err(bollard::errors::Error::from)?;
+    ///             // p.uri = uri.try_into().map_err(bollard::errors::Error::from)?;
```
[the remark this answers is not in the document]
Comment (izissise) [reply src/docker.rs:649]: Actually I wanted to show how to change the request, without actually modifying it so the test passes

**Comment by DDtKey** [comment]
Sorry, looks like my suggestion got missed by `rustfmt`

**Comment by izissise** [comment]
My bad I didn't check before end, also I'm on my work laptop without my github ssh key so I won't be able to fix it soon.

Have you tested the doc example? Does it compile? 

I'm wondering because I tried adding the `'static` lifetime but the example in the doc wouldn't compile https://github.com/fussybeaver/bollard/pull/459/commits/2c5c765265839658f1afc167e1dc0c2e87c411f3

**Comment by izissise** [comment]
Nevermind it just passed the test :) Thanks for the help :)

**Comment by izissise** [comment]
Also I wasn't careful and I commited the generated file

File: src/docker.rs [hunk 5]
```diff
@@ -90,6 +90,27 @@ pub(crate) enum ClientType {
...
+type CallbackTy = Box<dyn Fn(Request<BodyType>) -> CallbackRetTy + Send + Sync>;
+
+/// Callback transport trait
+pub trait CallbackTransport: Send + Sync {
+    /// Make a request, this returns a future
+    fn request(&self, request: Request<BodyType>) -> CallbackRetTy;
+}
+
+// auto impl for Fn(Request)
+impl CallbackTransport for CallbackTy {
```
[the remark this answers is not in the document]
Comment (izissise) [reply src/docker.rs]: Maybe `TransportTrait`? 
This would be `use bollard::TransportTrait`

File: src/docker.rs (hunk 3 above)
Comment (izissise) [reply src/docker.rs]: This probably mean that `BodyType` must become fully public

**Comment by izissise** [comment]
I cleaned the MR and it passes the doc test now :)

**Comment by izissise** [comment]
One last thing, is that using `Request<BodyType>` forces the users to depends on `hyper` directly.
There is multiple solution if we think that is a problem:
- re-export hyper from bollard
- Hide the type (BollardRequest)
- Something else?

**Comment by fussybeaver** [comment]
> One last thing, is that using `Request<BodyType>` forces the users to depends on `hyper` directly. There is multiple solution if we think that is a problem:
> 
>     * re-export hyper from bollard
> 
>     * Hide the type (BollardRequest)
> 
>     * Something else?

I think `hyper::Request` and `hyper::Response` are just aliases of `http::Request` and `http::Response`, so we can use those imports instead as the `http` crate is a more general purpose crate.

**Comment by izissise** [comment]
Yes `http` crate is better. my concern is adding an extra dependency in users `Cargo.toml`.
For example if I want to write a docker client using bollard, I just want to have bollard dependency. Without any changes I would at least need to know about `http` and bump version in the Cargo.toml when needed.

**Comment by fussybeaver** [comment]
Though that only applies if the user is using a custom request callback, at which point the user may wish to specify her own `http` library version...

I'm reading about some conflicting best practices about this when I read about it.. I'm wondering if someone more knowledgeable is available to briefly comment @jonhoo  ? :pray:

**Comment by DDtKey** [comment]
I believe it's a good practice to re-export types (or even whole crate) if it's supposed to be a public interface for the library.

I may want to have closure with explicit types definition (e.g: `|req: http::Request|`), so it's applicable not only for custom impl IMHO.

Another point here - versions should be consistent between client code and `bolard`. Let's say (unlikely), but tomorrow http2 is released, so users may want to upgrade but will have to use http1 until `bollard` is updated and etc.

File: src/docker.rs [hunk 6]
```diff
@@ -90,6 +90,29 @@ pub(crate) enum ClientType {
...
+    Custom {
+        scheme: String,
+    },
+}
+
+type TransportReturnTy =
+    Pin<Box<dyn Future<Output = Result<Response<hyper::body::Incoming>, Error>> + Send>>;
+
+/// Callback transport trait
+pub trait TransportTrait: Send + Sync {
```
Comment (DDtKey) [inline src/docker.rs:102]:
```suggestion
/// Callback transport trait
pub trait TransportTrait: Send + Sync {
```

a nit: I personally find such (pre/post)-fixes excessive, they don't have meaningful part - the type itself defines itself as a trait.
So it's subjective opinion, but `Transport` or `CustomTransport` should work well. Also `callback` in documentation looks a bit unrelated, it's just an interface to pass your custom transport implementation. 
It's just implemented for callback too

Comment (izissise) [reply src/docker.rs:102]: I'll make the changes

**Comment by fussybeaver** [comment]
> I may want to have closure with explicit types definition (e.g: `|req: http::Request|`, or even write a function), so it's applicable not only for custom impl IMHO.

It's not applicable for general usage of the library, outside of the use cases described here and associated tickets.

> Another point here - versions should be consistent between client code and `bollard`. Let's say (unlikely), but at the end of the day http v2 is released, so users may want to upgrade but will have to use http1 as well (in their deps) until `bollard` is updated and etc.

Maybe I don't understand exactly what you mean, but if http v2 is released, by re-exporting the http crate, the user is forced to use http v1 for the callback handler and he can't actually use http v2 inside bollard.

**Comment by DDtKey** [comment]
> Maybe I don't understand exactly what you mean, but if http v2 is released, by re-exporting the http crate, the user is forced to use http v1 for the callback handler and he can't actually use http v2 inside bollard.

No, what I mean is: user has its dependency to `http1` and it's used for something except `bollard`. e.g I have some custom code which I use in my application, and in addition to that I have custom implementation of transport for `bollard`. 
Let's say http2 is released and I've updated my code, but for `bollard` I'll need to have `http1` as part of my dependencies just because of custom implementation of the trait which requires me to use `http1`. 

But when types are re-exported, it's responsibility of bollard to update the version, not mine as a client. It's ofc still uses http1 under the hood and you will see this within Cargo.lock, but I don't have to manage this, since it's part of bollard.

And what can be useful here is that if types are re-exported, it's more obvious whether a crate update should be considered a breaking change. Otherwise, it's easier to update a dependency without realizing that some users depend on it.


> It's not applicable for general usage of the library, outside of the use cases described here and associated tickets.

It's still part of public api, it can be type alias or just re-export through some of modules - but it is, even for some particular use-case.

**Comment by fussybeaver** [comment]
OK, let's re-export these public types. I'm not entirely convinced ... but happy to defer to someone else's wisdom here and do some more reading.

**Comment by izissise** [comment]
LGTM

**Comment by fussybeaver** [comment]
Thanks for all the great work here!
